### PR TITLE
Properties: Add support for custom property types.

### PR DESCRIPTION
**Pull request by LlamaLad7**
Specified using `PropertyType.CUSTOM` and by setting `Property#customPropertyInfo`

### CODE REVIEW

**Review by Johni0702 — CHANGES_REQUESTED** [review]
Is it ok to use `KClass` as opposed to Java `Class` there? Are those usable in annotations from Java? And iirc we deprecated some methods because Kotlin reflection is slow to initialize, if these custom properties are enough to trigger its initialization, then we probably should refrain from using it here as well.

There's also a bunch of backwards-incompatible ABI changes (adding a new argument to a method with optional arguments is not backwards compatible, just like it wouldn't be in Java; though even with @JvmOverloads, it's not backwards compatible because of how default arguments are actually implemented).
Running `./gradlew :1.12.2:check` should yell at you (though that's not entirely your fault, previous commits have neglected to update the API file as well because CI didn't use to run `check`, ideally merge the essential-gradle-toolkit branch into this one, it fixes all of those), and running `./gradlew apiDump` should write the new ABI to the file, so you can commit expected changes.

Otherwise, the functionality itself seems good to me (without having tested it).

**Comment by LlamaLad7** [comment]
> Is it ok to use KClass as opposed to Java Class there? Are those usable in annotations from Java?

Yes, that's a weird Kotlin thing, the resultant bytecode only references `java.lang.Class`.

> And iirc we deprecated some methods because Kotlin reflection is slow to initialize

In Essential's commands thing I think we initialize it on a background thread. We could possibly do the same here?

**Comment by Johni0702** [comment]
> In Essential's commands thing I think we initialize it on a background thread. We could possibly do the same here?

I don't think we can. The properties are initialized in the constructor (and subclasses rely on that).
Or rather, it depends on whether simply having KClass instances is already enough to trigger that slow initialization. If it is, then we can't, and should probably switch to Class. If it isn't, and it's only the `constructors` call which initializes kotlin's reflection, then it's probably fine as is, cause that'll only be called when the GUI is opened, and by then kotlin's reflection will likely already have been initialized.

**Comment by LlamaLad7** [comment]
> it depends on whether simply having KClass instances is already enough to trigger that slow initialization. If it is, then we can't, and should probably switch to Class

From my testing, it seems having a KClass instance around triggers some initialization, though not as much as the `constructors` call, so I'll use `Class` instead, except in the annotation itself where Kotlin requires `KClass` as a facade for `Class`. (Similarly to `::class.java` literals, the intermediary KClass object is avoided when referencing `annotation.kClassProperty.java`.)

**Comment by LlamaLad7** [comment]
> adding a new argument to a method with optional arguments is not backwards compatible, just like it wouldn't be in Java; though even with @JvmOverloads, it's not backwards compatible because of how default arguments are actually implemented

Would you suggest adding an explicit overload instead? Or is there some more elegant solution.

File: src/main/kotlin/gg/essential/vigilance/Vigilant.kt [hunk 1]
```diff
@@ -407,7 +408,100 @@ abstract class Vigilant @JvmOverloads constructor(
                 placeholder = placeholder,
                 triggerActionOnInitialization = triggerActionOnInitialization,
                 hidden = hidden,
-                action = action
+                action = action,
+            )
+        }
+
+        // Same as [property] but does not need to remain binary-compatible.
```
Comment (Johni0702) [inline src/main/kotlin/gg/essential/vigilance/Vigilant.kt:415]: `property` should probably delegate to this one then, so we don't have the same code in two places.

File: src/main/kotlin/gg/essential/vigilance/data/Categories.kt [hunk 2]
```diff
@@ -57,6 +57,15 @@ class PropertyItem(val data: PropertyData, val subcategory: String) : CategoryIt
...
             )
             PropertyType.BUTTON -> ButtonComponent(data.attributesExt.placeholder, data)
+            PropertyType.CUSTOM -> {
+                val propertyInfoClass = data.attributesExt.customPropertyInfo
+                (propertyInfoClass
+                    .constructors
+                    .find { it.parameters.isEmpty() }
+                    ?.newInstance() as? PropertyInfo)
+                    ?.createSettingComponent(data.getValue())
+                    ?: error("PropertyInfo class $propertyInfoClass has no zero-arg constructor!")
```
Comment (Johni0702) [inline src/main/kotlin/gg/essential/vigilance/data/Categories.kt:67]:
```suggestion
                propertyInfoClass
                    .getConstructor()
                    .newInstance()
                    .createSettingComponent(data.getValue())
```
Assuming the error thrown by `getConstructor(args...)` is at least as useful as the one we construct manually.

**Review by Johni0702 — CHANGES_REQUESTED** [review]

**Review by Johni0702 — APPROVED** [review]

**Pull request merged**